### ISSUE TITLE
Allow autoSave and formatOnSave

### ISSUE THREAD

**Issue by andrecasal** [microsoft/vscode#45997]
[VS Code documentation](https://code.visualstudio.com/docs/getstarted/settings) regarding configuration says that formatOnSave only works when autoSave is turned off.

The formatOnSave config:
```
// Format a file on save. A formatter must be available, the file must not be auto-saved,
// and editor must not be shutting down.
  "editor.formatOnSave": true,
```
and the autoSave config:
```
// Controls auto save of dirty files. Accepted values:  'off', 'afterDelay',
// 'onFocusChange' (editor loses focus), 'onWindowChange' (window loses focus).
// If set to 'afterDelay', you can configure the delay in 'files.autoSaveDelay'.
  "files.autoSave": "off",

// Controls the delay in ms after which a dirty file is saved automatically.
// Only applies when 'files.autoSave' is set to 'afterDelay'
  "files.autoSaveDelay": 1000,
```
Both live inside the user settings.

It would be a great developer experience to have files automatically saved AND format the documents automatically.

There's a closed issue with a similar title [#37820](https://github.com/Microsoft/vscode/issues/37820), but the author was looking to solve an extension bug and it got closed when it was solved. I think this issue remains an important feature request.

**Comment by campelo** [comment]
>I think this issue remains an important feature request.

I agree with you.

**Comment by eladcandroid** [comment]
Any news about this issue? It can be very helpful...

**Comment by varthc** [comment]
+1 This feature would be awesome.

**Comment by ildarmgt** [comment]
It doesn't say it for `"files.trimTrailingWhitespace": true`, but it doesn't seem to trim trailing white spaces  when it auto saves `"files.autoSave": "afterDelay"`, but works when you save the file (like with ctrl-s). Similar behavior seen elsewhere. `Autosave` treated differently than regular `Save` on my end.

**Comment by andraz** [comment]
I would paraphrase the request a bit:

How can we get the functionality a setting named `"editor.formatOnAutoSave": true` would provide us?

At the moment I am struggling to make Prettier format code as soon as I change focus which triggers auto-save. Both work on their own, but do not work together.

expected result:
File gets formatted first, then gets saved.

actual result:
File gets saved without formatting.

**Comment by JerryGreen** [comment]
> expected result:
> File gets formatted first, then gets saved.
> 
> actual result:
> File gets saved without formatting.

This is not true. Having `autoSave` and `formatOnSave` both set to true, it works properly: it gets formatted, then saved.

_However_, I agree that `"editor.formatOnAutoSave": true | false` setting is a good idea. I don't want my file to be formatted on `autoSave`, I want it to get formatted on a manual save only.

If you try to make a keybinding `cmd+s` to assign `editor.action.format`, then yes - it works improperly, it gets saved, then gets formatted.

**Comment by lopsch** [comment]
> > expected result:
> > File gets formatted first, then gets saved.
>>
> > actual result:
> > File gets saved without formatting.
> 
> This is not true. Having `autoSave` and `formatOnSave` both set to true, it works properly: it gets formatted, then saved.

`autoSave` can't be set to `true`. If it is set to `afterDelay` the `formatOnSave` will not work even if this is set to `true`:

> Format a file on save. A formatter must be available, the file must not be saved after delay, and the editor must not be shutting down.

At least this is my understanding.

**Comment by JerryGreen** [comment]
@lopsch oh sorry by `autoSave` set to `true` I actually meant it's set to `onFocusChange`

What I have, basically, in `settings.json`:

```json
"editor.codeActionsOnSave": {
  "source.fixAll.tslint": true
},
"editor.formatOnSave": true,
"files.autoSave": "onFocusChange",
```

And it works just fine. At least from VSCode side. Sometimes it happened that `tslint` was broken in newer versions, so I had to change version to an older one (currently 1.2.1 TSLint for VSCode works just fine).

So I don't see any problem here, actually. And...

> It would be a great developer experience to have files automatically saved AND format the documents automatically.

It's already working in such a way. Both [Prettier](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) (which is controller by `editor.formatOnSave`) and [TSLint](https://marketplace.visualstudio.com/items?itemName=ms-vscode.vscode-typescript-tslint-plugin) (which is controller by `editor.codeActionsOnSave` and `source.fixAll.tslint`). At least they are working for me, automatically, on save, which is happened when the editor loose focus (or if I save manually too).

So I think this issue should be closed.

**Comment by andrecasal** [comment]
This feature isn't yet implemented @JerryGreen 

To reproduce:
1. Copy this to an HTML file:
```
<!DOCTYPE html>
<html>
<head>
	<title>Hello World</title>
</head>
<body>
	<div>Hello world!</div>
</body>
</html>
```
2. Set the following settings.json settings:
```
{
	"html.format.indentInnerHtml": true,
	"files.autoSave": "afterDelay"
}
```
3. Make a change to the file and wait for the afterDelay to trigger a save.
4. Notice that the `<head>` and the `<body>` tags are not indented as per the `"html.format.indentInnerHtml": true` setting.

However, if you save the file manually, it will trigger the correct indentation.

**Comment by davidbarton** [comment]
For me it is very important part of developer experience. I know it is already possible to hack some linters to get the code formatted for you. For example Prettier + TSLint or rust-lang can do that. As described above this should be possible for all formatters using simple setting. Is there a chance to get official response from team on this issue please? I see @jrieken is now assigned to the issue.

**Comment by carlogonzalo** [comment]
+1 for this feature and suprising that it is not possible at the moment considering awsome vscode.

**Comment by rtbo** [comment]
+1
I confirm that auto formatting is done with `onFocusChange` and `onWindowChange`, but not with `afterDelay`.

Note that some formatters take a substantial time to execute and editing experience could quickly become annoying, so the current behavior is not unsound.
An additional setting like `editor.formatOnAutoSave` is needed.

**Comment by symonxdd** [comment]
Yup, would be nice to have this feature, 2 years later...

**Comment by julix-unity** [comment]
"We are happy to listen and to reconsider." - Please reconsider. Fair amount of up votes. What's the main concern with implementing this? Performance?

**Comment by steinybot** [comment]
What's worse is that if you manually save after the auto save delay then it doesn't run the formatter either (#119857).

**Comment by maljaroudi** [comment]
I don't see why we still don't have this option in vscode. If the concern on implementing this is about performance, then that's why it's an option, like any other performance impacting options.

**Comment by skngetich** [comment]
How can we get it to be added as a feature? 
I have been waiting on this feature for 3 years now

**Comment by RafaelGlober** [comment]
I literally discover the "AutoSave" setting today, and instantly I started searching "how to format on autosave" and now I'm commenting on this thread... so if you can see... both features "auto save" and "format on auto save" are very needed.

**Comment by ktvdev** [comment]
In the same boat here, somehow missed that autosave was a thing, discovered it is, tried to make it work with format on save and no go. 

I know very little about the inner-workings of VSCode since I've had little interest in that aspect of it so I'm not sure if a better workaround exists, but in the meantime the quickest solution I've found is using an extension that forces the format command.

The extension in question: [Run on Save](https://marketplace.visualstudio.com/items?itemName=pucelle.run-on-save)
Specifically the one I linked, the first one I tried had more downloads but didn't work for me at all.

<br>

&nbsp; &nbsp;Then in your settings.json add something along the lines of:

```
"runOnSave.commands": [
    {
        "match": ".ps1",
        "command": "editor.action.formatDocument",
        "runIn": "vscode"
    }
]

```

<br>

Enable `files.autoSave` as normal, match your filetype or use your own regex in settings and you should be good to go.

**Comment by rudyhadoux** [comment]
Hi, "format on auto save after delay" is a real need indeed.

**Comment by skngetich** [comment]
> In the same boat here, somehow missed that autosave was a thing, discovered it is, tried to make it work with format on save and no go.
> 
> I know very little about the inner-workings of VSCode since I've had little interest in that aspect of it so I'm not sure if a better workaround exists, but in the meantime the quickest solution I've found is using an extension that forces the format command.
> 
> The extension in question: [Run on Save](https://marketplace.visualstudio.com/items?itemName=pucelle.run-on-save)
> Specifically the one I linked, the first one I tried had more downloads but didn't work for me at all.
> 
> 
>    Then in your settings.json add something along the lines of:
> 
> ```
> "runOnSave.commands": [
>     {
>         "match": ".ps1",
>         "command": "editor.action.formatDocument",
>         "runIn": "vscode"
>     }
> ]
> ```
> 
> 
> Enable `files.autoSave` as normal, match your filetype or use your own regex in settings and you should be good to go.

As we wait for this feature.This hack by @ktvdev works

**Comment by feinlord** [comment]
> В той же лодке здесь, как-то пропустил, что автосохранение было чем-то особенным, обнаружил, что это так, попытался заставить его работать с форматом при сохранении и ничего не делать.
> 
> Я очень мало знаю о внутренней работе VSCode, так как я мало интересовался этим аспектом, поэтому я не уверен, существует ли лучший обходной путь, но пока что самое быстрое решение, которое я нашел, - это использование расширения. это заставляет команду форматирования.
> 
> Рассматриваемое расширение: « [Запускать при сохранении». В](https://marketplace.visualstudio.com/items?itemName=pucelle.run-on-save)
> частности, то, которое я связал, первое, которое я попробовал, имело больше загрузок, но у меня не работало.
> 
> 
>    Затем в вашем settings.json добавьте что-нибудь вроде:
> 
> ```
> "runOnSave.commands": [
>     {
>         "match": ".ps1",
>         "command": "editor.action.formatDocument",
>         "runIn": "vscode"
>     }
> ]
> ```
> 
> 
> Включите `files.autoSave`как обычно, сопоставьте свой тип файла или используйте собственное регулярное выражение >в настройках, и все будет в порядке.

Doesn't work for me

**Comment by ktvdev** [comment]
> > В той же лодке здесь, как-то пропустил, что автосохранение было чем-то особенным, обнаружил, что это так, попытался заставить его работать с форматом при сохранении и ничего не делать.
> > Я очень мало знаю о внутренней работе VSCode, так как я мало интересовался этим аспектом, поэтому я не уверен, существует ли лучший обходной путь, но пока что самое быстрое решение, которое я нашел, - это использование расширения. это заставляет команду форматирования.
> > Рассматриваемое расширение: « [Запускать при сохранении». В](https://marketplace.visualstudio.com/items?itemName=pucelle.run-on-save)
> > частности, то, которое я связал, первое, которое я попробовал, имело больше загрузок, но у меня не работало.
> >    Затем в вашем settings.json добавьте что-нибудь вроде:
> > ```
> > "runOnSave.commands": [
> >     {
> >         "match": ".ps1",
> >         "command": "editor.action.formatDocument",
> >         "runIn": "vscode"
> >     }
> > ]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Включите `files.autoSave`как обычно, сопоставьте свой тип файла или используйте собственное регулярное выражение >в настройках, и все будет в порядке.
> 
> Doesn't work for me

I would check your extensions output log in VSCode and see if Save on Run is showing any errors. Still works for me as of VSCode v1.57.1

**Comment by jamie9762** [comment]
> In the same boat here, somehow missed that autosave was a thing, discovered it is, tried to make it work with format on save and no go.
> 
> I know very little about the inner-workings of VSCode since I've had little interest in that aspect of it so I'm not sure if a better workaround exists, but in the meantime the quickest solution I've found is using an extension that forces the format command.
> 
> The extension in question: [Run on Save](https://marketplace.visualstudio.com/items?itemName=pucelle.run-on-save)
> Specifically the one I linked, the first one I tried had more downloads but didn't work for me at all.
> 
> 
>    Then in your settings.json add something along the lines of:
> 
> ```
> "runOnSave.commands": [
>     {
>         "match": ".ps1",
>         "command": "editor.action.formatDocument",
>         "runIn": "vscode"
>     }
> ]
> ```
> 
> 
> Enable `files.autoSave` as normal, match your filetype or use your own regex in settings and you should be good to go.

it's work. thanks!

**Comment by StefanIGit** [comment]
I spent like 1 day figuring out that it does not work as expected. My problem with this `afterDelay` is that it is not documented in a way you know this does not trigger auto format. 
At least add a warning/hint to the docu https://code.visualstudio.com/docs/editor/codebasics#_save-auto-save

**Comment by manoharreddyporeddy** [comment]
After an hour of struggle... below steps worked. Full details below.

1. Install this extension

https://marketplace.visualstudio.com/items?itemName=pucelle.run-on-save

2. Add below json, to below file:

File:

    <your-project-directory>\.vscode\settings.json
        OR
    %UserProfile%\AppData\Roaming\Code\User\settings.json

JSON:

NOTE: Make sure commas before and after below block.

```json
    "files.autoSave": "afterDelay",
    "files.autoSaveDelay": 1000,
    "runOnSave.statusMessageTimeout": 3000,
    "runOnSave.commands": [
        {
            "match": "*.*",
            "command": "git add *",
            "runIn": "terminal"
        }
    ],
```

**Comment by granda17** [comment]
> I've found an extension that works like a charm with
> 
> ```json
> {
>     "files.autoSave": "afterDelay",
>     "editor.formatOnSave": true
> }
> ```
> 
> Just install it and you're good to go!
> 
> [Format on auto save extension](https://marketplace.visualstudio.com/items?itemName=BdSoftware.format-on-auto-save) by BDSoftware

There's not a lot of info on the extension page.  Other than the above config items, are there any that need to be defaulted?  I tried the extension, and it did not work, but I've been trying so many different things to get this functionality over the last day or so, I'm not sure if I've got something configured that shouldn't be.

Going to try the "run-on-save" option now.

**Comment by claired4l** [comment]
I've just used auto-save with `onFocusChange` and not `afterDelay` as @jerrygreen suggested and it works like a charm 😃

**Comment by BrenoJesusFernandes** [comment]
> In the same boat here, somehow missed that autosave was a thing, discovered it is, tried to make it work with format on save and no go.
> 
> I know very little about the inner-workings of VSCode since I've had little interest in that aspect of it so I'm not sure if a better workaround exists, but in the meantime the quickest solution I've found is using an extension that forces the format command.
> 
> The extension in question: [Run on Save](https://marketplace.visualstudio.com/items?itemName=pucelle.run-on-save) Specifically the one I linked, the first one I tried had more downloads but didn't work for me at all.
> 
> 
>    Then in your settings.json add something along the lines of:
> 
> ```
> "runOnSave.commands": [
>     {
>         "match": ".ps1",
>         "command": "editor.action.formatDocument",
>         "runIn": "vscode"
>     }
> ]
> ```
> 
> 
> Enable `files.autoSave` as normal, match your filetype or use your own regex in settings and you should be good to go.

Thx this actually works for me <3

**Comment by BrenoJesusFernandes** [comment]
It's a shame we do not have this as native feature

**Comment by vec715** [comment]
any progress? I'm switching to Goland for this feature lol

**Comment by dexbyte** [comment]
I realized that changing the settings of auto-save from `afterDelay` to `onFocusChange` is a better choice since I typically would want it to save when my focus change, it also stops the server from restarting repetitively on small changes and formatting works as well.

**Comment by MarceloViannaDev** [comment]
> I've found an extension that works like a charm with
> 
> ```json
> {
>     "files.autoSave": "afterDelay",
>     "editor.formatOnSave": true
> }
> ```
> 
> Just install it and you're good to go!
> 
> [Format on auto save extension](https://marketplace.visualstudio.com/items?itemName=BdSoftware.format-on-auto-save) by BDSoftware

The hero of the heroes!!! 🤘

**Comment by Scott-McMullan-ABB** [comment]
Adding my support for this feature. Although [Run On Save](https://marketplace.visualstudio.com/items?itemName=pucelle.run-on-save) solves the issue, this should be a native feature.
 
Honestly, the running commands on save should in itself be native.

**Comment by GiancarloFusiello** [comment]
Also adding my support for this feature.
